learningequality/kolibri
# Device Permissions - p2p sync Super Admin

### Summary


**Features**

- Updates the Device > Permissions page to include filters for "Permissions", "User Type" and, when there are multiple facilities, "Facility".
- Device > Permissions table now shows all users on the device.
- Device > Permissions filter values are persisted between the "View/Edit Permissions" page and going back to the table.
- Adds "Facility" column to the Device > Permissions table when there are multiple facilities.

**PaginatedListContainer Change**

`PaginatedListContainer` has a slot called `otherFilter` which was designed to give space for a single dropdown filter set left of the search filter input. 

This PR requires 2-3 filters to be inserted there.

So I just updated the `KGridItem` spans in that component to fit in a way that (IMO):

- Doesn't make other places where `PaginatedListContainer` is used look bad.
- Does make this feature work and look as expected.

The other place I noted this is Facility > Users:

__BEFORE__

![original-facility-users-table-layout](https://user-images.githubusercontent.com/6356129/81871813-6c398f80-952d-11ea-8984-1de2c0fa402d.png)

__AFTER__

![facility-users-table-with-variant-filter-layout](https://user-images.githubusercontent.com/6356129/81871842-7b204200-952d-11ea-8dd4-c0b4248ef5d6.png)

Would love some feedback on this change @jonboiser @jtamiace 


### Reviewer guidance


**SETUP**

You need to do some setup and need the following:

- A `KOLIBRI_HOME` folder with one facility 
- A `KOLIBRI_HOME` folder with **more than one** facility
- Both of those need one or more of the following user types created in each facility:

> Learner
> Facility Coach
> Class Coach
> Admin

**HOW TO TEST**

__Universal things to test:__

- `Device > Permissions` table in general.
- Pagination - can you navigate between pages. Is the table properly reset after changing filters after having navigated into table-pages? Unfortunately, you'll need > 30 users to test any of this - or you can take my word for it that it works :-p
- Filters functioning as expected. Set users to 'Can manage content' by going to `Device > Permissions > Edit permissions` and setting a user to `Can manage content`. Do the same with `Super admin`. Make sure that if you set a Learner to be a `Super admin` that they are represented as such in the table and appear when you filter for `Super admins`.
- If you set filters and then navigate to "Edit Permissions" or "View Permissions" - when you click the back button on your browser, your filters are still there and are applied. Should also persist if you use the `<-` button in the AppBar.

__Multiple Facility vs Single Facility__

- The `Facility` column on `Device > Permissions` should only be visible on multiple-facility device.
- The `Facility` filter should only show when on a multiple-facility device.


### References


[Notion Specs](https://www.notion.so/learningequality/Superusers-can-view-all-Users-in-Device-Permissions-page-8-pts-0-14-misc-UC3-4-169d4360ff7b47f7a4de4afd0b6b32a0)

[Figma](https://www.figma.com/file/LM0YUkbWAOIerZ3EEg0FN6bS/0.14-P2P-Sync?node-id=605%3A1769)

----

### Contributor Checklist


PR process:


Testing:


### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md
